internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -90,33 +90,29 @@ object ClassToAPI {
    *
    * In order to fully fix this issue, the Scala compiler should do the desired behaviour
    * in 2.13.x. 2.12.x is already lost since this fix is binary incompatible. The appropriate
-   * JIRA ticket is here: https://github.com/scala/bug/issues/2034.
+   * issue is here: https://github.com/scala/bug/issues/2034.
    *
-   * @return The canonical name if not null, the full name otherwise.
+   * @return The canonical name if not null, the blank string otherwise.
    */
-  def handleMalformedNameOf(c: Class[_], isRecursive: Boolean = false): String = {
+  def classCanonicalName(c: Class[_]): String = {
     if (c == null) "" // Return nothing if it hits the top-level class
     else {
       val className = c.getName
-      // Adds a `dollar` if it's a recursive call, else nothing
-      val atEnd: String = if (isRecursive) "$" else ""
       try {
         val canonicalName = c.getCanonicalName
         if (canonicalName == null) className
-        else canonicalName + atEnd
+        else canonicalName
       } catch {
         case malformedError: java.lang.InternalError
             if malformedError.getMessage.contains("Malformed class name") =>
           val enclosingClass = c.getEnclosingClass
           val enclosingName = enclosingClass.getName
           val restOfName = c.getName.stripPrefix(enclosingName)
-          handleMalformedNameOf(enclosingClass, true) + restOfName + atEnd
+          classCanonicalName(enclosingClass).stripSuffix("$") + "." + restOfName
```

Comment:
> Then your concern is of the status quo, not specifically this proposed change. Because currently (as demonstrated in the spec as of #380) you can't tell the difference between:
  
Not really. Zinc assumes that both class and companion class have the same name. This is an invariant embedded in all the logic that handles incrementality. It not only happens in `getCanonicalClass`, but also in the bridge, where the names for both are the same.
  
This change, however, creates a clash because two different class files (one with a valid name and another one with an invalid name) will map to the same name. Those don't have any relationship between them, as in the case of class and companion class.
  
I'm not saying this patch is wrong. I like the fact that the notation is more Scala-like. But we must first understand the consequences of such a change.